File: FLEX.Data.Oracle/QueryExecutorBase.vb
Imports Dapper
Imports Finsa.Caravan.Extensions
Imports FLEX.Common.Data

Public MustInherit Class QueryExecutorBase
   Implements IQueryExecutor

   Public Function FillDataTableFromQuery(query As String) As DataTable Implements IQueryExecutor.FillDataTableFromQuery
      Using connection = OpenConnection()
         Return connection.Query(query).ToDataTable()
      End Using
   End Function

   Public Function FillDataTableFromQuery(query As String, parameters As Object) As DataTable Implements IQueryExecutor.FillDataTableFromQuery
      Using connection = OpenConnection()
         Return connection.Query(query, parameters).ToDataTable()
      End Using
   End Function

   Public MustOverride Sub ElaborateConnectionString(ByRef connectionString As String) Implements IQueryExecutor.ElaborateConnectionString

   Public MustOverride Function OpenConnection() As IDbConnection Implements IQueryExecutor.OpenConnection
End Class

File: Caravan.DataAccess/Core/QueryExecutorBase.cs
using System.Data;
using System.Data.Common;
using Dapper;
using Finsa.Caravan.Extensions;

namespace Finsa.Caravan.DataAccess.Core
{
    public abstract class QueryExecutorBase : IQueryExecutor
    {
       public abstract DatabaseVendor Vendor { get; }

       public DataTable FillDataTableFromQuery(string query)
        {
            using (var connection = OpenConnection())
            {
                return connection.Query(query).ToDataTable();
            }
        }

        public DataTable FillDataTableFromQuery(string query, object parameters)
        {
            using (var connection = OpenConnection())
            {
                return connection.Query(query, parameters).ToDataTable();
            }
        }

        public abstract void ElaborateConnectionString(ref string connectionString);

        public abstract DbConnection OpenConnection();
    }
}